### PR TITLE
Add endpoint to auto-assign people in leagues

### PR DESCRIPTION
**Pull request by varniethan**
Initial attempt for https://trello.com/c/seZoenpX/87-add-endpoint-for-auto-assigning-people-from-leagues-make-sure-to-add-round-integration.

### CODE REVIEW

**Review by WilfSilver — CHANGES_REQUESTED** [review]
This is not building! Check the actions

File: packages/types/src/league.ts [hunk 1]
```diff
@@ -1,10 +1,18 @@
 import type { Query } from "./db_client.js";
 import type { MongoDBItem, ObjectId, Sport } from "./utils.js";
-
+import type { AbilityLevel, User } from "@esp-group-one/types";
```
Comment (WilfSilver) [inline packages/types/src/league.ts:3]: AHHHHH SPACE PLEASE!

File: packages/types/src/league.ts [hunk 2]
```diff
@@ -1,10 +1,18 @@
...
 interface BaseLeague extends MongoDBItem {
   name: string;
   ownerIds: ObjectId[];
   sport: Sport;
+  round: number;
+  people: User[];
```
Comment (WilfSilver) [inline packages/types/src/league.ts:9]: This is not required, nor do we want to be storing duplicate information. Each user stores the `leagues` which they are apart of

File: packages/types/src/league.ts [hunk 3]
```diff
@@ -1,10 +1,18 @@
...
+}
+
+export interface MatchProposal {
+  ownerId: ObjectId;
+  players: ObjectId[];
+  abilityLevel: AbilityLevel;
 }
```
Comment (WilfSilver) [inline packages/types/src/league.ts:16]: Why? We already have a `MatchProposal` type.

File: packages/types/src/league.ts [hunk 4]
```diff
@@ -22,12 +30,16 @@ export interface LeagueCreation {
   name: string;
   sport: Sport;
   private: boolean;
+  round: number;
+  people: User[];
```
Comment (WilfSilver) [inline packages/types/src/league.ts:34]:
```suggestion
```

File: packages/types/src/league.ts [hunk 5]
```diff
@@ -36,10 +48,13 @@ export function censorLeague(league: League): CensoredLeague {
     name: league.name,
     sport: league.sport,
     private: league.private,
+    round: league.round,
+    people: league.people,
```
Comment (WilfSilver) [inline packages/types/src/league.ts:52]:
```suggestion
```

File: packages/types/src/league.ts [hunk 6]
```diff
@@ -22,12 +30,16 @@ export interface LeagueCreation {
...
+  round: number;
+  people: User[];
 }
 
 export interface CensoredLeague extends MongoDBItem {
   name: string;
   sport: Sport;
   private: boolean;
+  round: number;
+  people: User[];
```
Comment (WilfSilver) [inline packages/types/src/league.ts:42]:
```suggestion
```

File: packages/backend/src/league/league_controller.ts
Comment (WilfSilver) [inline packages/backend/src/league/league_controller.ts:1]: `creation` needs to be updated to initialise the league at `round` 0

File: packages/backend/src/league/league_controller.ts [hunk 7]
```diff
@@ -265,4 +268,62 @@ export class LeaguesController extends ControllerWrap<League> {
...
+      for (const person of league.people as User[]) {
+        peopleByAbilityLevel[person.abilityLevel].push(person._id);
+      }
+
+      // Generate match proposals for each ability level
+      for (const abilityLevel of [
+        "beginner",
+        "intermediate",
+        "advanced",
+      ] as const) {
```
Comment (WilfSilver) [inline packages/backend/src/league/league_controller.ts:310]: Please iterate over the enum `AbilityLevel`, therefore `as const` won't be required

File: packages/backend/src/league/league_controller.ts [hunk 8]
```diff
@@ -265,4 +268,62 @@ export class LeaguesController extends ControllerWrap<League> {
...
+
+      const matchProposals: MatchProposal[] = [];
+
+      // Group people by ability level
+      const peopleByAbilityLevel: Record<AbilityLevel, ObjectId[]> = {
+        beginner: [],
+        intermediate: [],
+        advanced: [],
+      };
+      for (const person of league.people as User[]) {
```
Comment (WilfSilver) [inline packages/backend/src/league/league_controller.ts:301]: You will have to do a query to find the people in the league:

```ts
await (await db.users()).find({ leagues: league._id })
```

File: packages/backend/src/league/league_controller.ts [hunk 9]
```diff
@@ -265,4 +268,62 @@ export class LeaguesController extends ControllerWrap<League> {
...
+        peopleByAbilityLevel[person.abilityLevel].push(person._id);
```
Comment (WilfSilver) [inline packages/backend/src/league/league_controller.ts:302]: `abilityLevel` does not appear on the person as it is dependent on the sport. (See `User.sports`)

File: packages/backend/src/league/league_controller.ts [hunk 10]
```diff
@@ -265,4 +268,62 @@ export class LeaguesController extends ControllerWrap<League> {
...
+      }
+
+      const matchProposals: MatchProposal[] = [];
+
+      // Group people by ability level
+      const peopleByAbilityLevel: Record<AbilityLevel, ObjectId[]> = {
+        beginner: [],
+        intermediate: [],
+        advanced: [],
+      };
```
Comment (WilfSilver) [inline packages/backend/src/league/league_controller.ts:300]: please leave more blank lines

**Review by WilfSilver — CHANGES_REQUESTED** [review]
Looks good, however you seem to have made quite significant errors in the assumptions you have made about our current data types + database structure.

I would recommend sitting down and spending like 30 minutes to an hour just trying to figure out the structure of the database and why our types are the way they are.

Github copilot does not understand any of that as it doesn't look at the lsp server! It is just predicting the next couple characters based off of all the code in the world, so for things like this, it really is not very good (as it is just guessing the types). Therefore you need to understand the code base for it instead of thinking "oh yeah let's just commit this and see what happens" as it just wastes time (which is the complete opposite of what copilot is there for.)

File: packages/backend/src/league/league_controller.ts (hunk 7 above)
Comment (varniethan) [reply packages/backend/src/league/league_controller.ts:310]: do you mean: for (const abilityLevel in AbilityLevel) {
  const people = peopleByAbilityLevel[abilityLevel as keyof typeof AbilityLevel];
  // ...
}

Comment (WilfSilver) [reply packages/backend/src/league/league_controller.ts]: Yup you can apparently do:

```ts
Object.fromEntries(
  Object
    .keys(AbilityLevel)
    .map((k) => [k, []])
)
```

Note here, we are using `Object.keys` which gives you the enum basically as an array: `["beginner"...]` and converting it to a tuple of `[string, unknown[]]`, which is the form `Object.fromEntries` takes to create an object with keys that are located in index `0` (so the strings) and values in index `1` (so the `[]`).

You will also have to specify the type which as you've noted is

```ts
Record<keyof typeof AbilityLevel, ObjectId[]>
```

Comment (WilfSilver) [reply packages/backend/src/league/league_controller.ts]: Oh oops I realised I have given you the code for https://github.com/ESP-Number-One/MatchyoSports/pull/91#discussion_r1521676804 and not this comment

You code there should work perfectly fine

**Review by WilfSilver — APPROVED** [review]
I think I'm loosing it, hopefully everything is covered, if not we will fix later

**Pull request merged**